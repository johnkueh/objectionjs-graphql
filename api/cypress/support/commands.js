Cypress.Commands.add('deleteUser', email => {
  return cy.task('deleteUser', email);
});

Cypress.Commands.add('factory', args => {
  return cy.task('factory', args);
});

Cypress.Commands.add('createUserAndLogin', args => {
  cy.task('factory', {
    method: 'create',
    type: 'user',
    args,
    include: ['jwt']
  }).then(user => {
    cy.setCookie('jwt', user.jwt);
    cy.wrap(user);
  });
});

Cypress.Commands.add('createUserWithWorkspacesAndLogin', args => {
  cy.task('factory', {
    method: 'create',
    type: 'userWithWorkspace',
    args,
    include: ['jwt', 'workspaces']
  }).then(user => {
    cy.setCookie('jwt', user.jwt);
    cy.wrap(user);
  });
});
