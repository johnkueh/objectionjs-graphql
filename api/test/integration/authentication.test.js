import factory from 'factory-girl';
import '../support/transactional-tests';
import '../factories';
import { query } from '../support/apollo-test-helper';

describe('Logging in', () => {
  const LOGIN = `
    mutation($input: LoginInput!) {
      login(input: $input) {
        jwt
        user {
          id
          name
          email
        }
      }
    }
  `;
  beforeEach(async () => {
    await factory.create('user', {
      email: 'john@doe.com',
      password: 'password'
    });
  });

  it('is not able to login with wrong credentials', async () => {
    const res = await query({
      query: LOGIN,
      variables: {
        input: {
          email: 'john@doe.com',
          password: 'wrongpassword'
        }
      }
    });

    expect(res.errors[0].extensions.exception.errors).toEqual({
      auth: 'Please check your credentials and try again.'
    });
  });

  it('is able to login with correct credentials', async () => {
    const res = await query({
      query: LOGIN,
      variables: {
        input: {
          email: 'john@doe.com',
          password: 'password'
        }
      }
    });

    expect(res.data.login).toEqual({
      jwt: expect.any(String),
      user: {
        id: expect.any(String),
        email: 'john@doe.com',
        name: 'John Doe'
      }
    });
  });
});

describe('Signing up', () => {
  const SIGNUP = `
    mutation($input: SignupInput!) {
      signup(input: $input) {
        jwt
        user {
          id
          name
          email
        }
      }
    }
  `;
  it('is able to signup successfully', async () => {
    const res = await query({
      query: SIGNUP,
      variables: {
        input: {
          name: 'John Doe',
          email: 'john@doe.com',
          password: 'wrongpassword'
        }
      }
    });

    expect(res.data.signup).toEqual({
      jwt: expect.any(String),
      user: {
        id: expect.any(String),
        email: 'john@doe.com',
        name: 'John Doe'
      }
    });
  });

  it('is not able to signup with missing fields', async () => {
    const res = await query({
      query: SIGNUP,
      variables: {
        input: {
          name: '',
          email: 'hello@tes',
          password: 'short'
        }
      }
    });

    expect(res.errors[0].extensions.exception.errors).toEqual({
      name: 'Name must be at least 1 characters',
      email: 'Email must be a valid email',
      password: 'Password must be at least 6 characters'
    });
  });

  it('is not able to signup with taken email', async () => {
    await factory.create('user', {
      email: 'john@doe.com',
      password: 'password'
    });

    const res = await query({
      query: SIGNUP,
      variables: {
        input: {
          name: 'John Doe',
          email: 'john@doe.com',
          password: 'password'
        }
      }
    });

    expect(res.errors[0].extensions.exception.errors).toEqual({
      email: 'Email is already taken'
    });
  });
});
