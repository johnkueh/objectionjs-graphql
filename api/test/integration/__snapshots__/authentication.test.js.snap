// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Logging in is able to login with correct credentials 1`] = `
Object {
  "data": Object {
    "login": Object {
      "jwt": Any<String>,
      "user": Object {
        "email": "john@doe.com",
        "id": Any<String>,
        "name": "John Doe",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Logging in is not able to login with wrong credentials 1`] = `
Object {
  "code": "BAD_USER_INPUT",
  "exception": Object {
    "errors": Object {
      "auth": "Please check your credentials and try again.",
    },
  },
}
`;

exports[`Signing up is able to signup successfully 1`] = `
Object {
  "data": Object {
    "signup": Object {
      "jwt": Any<String>,
      "user": Object {
        "email": "john@doe.com",
        "id": Any<String>,
        "name": "John Doe",
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Signing up is not able to signup with missing fields 1`] = `
Object {
  "code": "BAD_USER_INPUT",
  "exception": Object {
    "errors": Object {
      "email": "Email must be a valid email",
      "name": "Name must be at least 1 characters",
      "password": "Password must be at least 6 characters",
    },
  },
}
`;

exports[`Signing up is not able to signup with taken email 1`] = `
Object {
  "code": "BAD_USER_INPUT",
  "exception": Object {
    "errors": Object {
      "email": "Email is already taken",
    },
  },
}
`;
